pychiquito/dsl.py

```diff
@@ -2,83 +2,159 @@
+    def shared(self: Circuit, name: str) -> Shared:
+        assert self.mode == CircuitMode.SETUP
+        return Shared(self.ast.add_shared(name, 0), 0)
 
-    def shared_with_phase(self: CircuitContext, name: str, phase: int) -> Shared:
-        return Shared(self.circuit.add_shared(name, phase), 0)
+    def shared_with_phase(self: Circuit, name: str, phase: int) -> Shared:
+        assert self.mode == CircuitMode.SETUP
+        return Shared(self.ast.add_shared(name, phase), 0)
 
-    def fixed(self: CircuitContext, name: str) -> Fixed:
-        return Fixed(self.circuit.add_fixed(name), 0)
+    def fixed(self: Circuit, name: str) -> Fixed:
+        assert self.mode == CircuitMode.SETUP
+        return Fixed(self.ast.add_fixed(name), 0)
 
-    def expose(self: CircuitContext, signal: Queriable, offset: ExposeOffset):
+    def expose(self: Circuit, signal: Queriable, offset: ExposeOffset):
+        assert self.mode == CircuitMode.SETUP
         if isinstance(signal, (Forward, Shared)):
-            self.circuit.expose(signal, offset)
+            self.ast.expose(signal, offset)
         else:
             raise TypeError(f"Can only expose ForwardSignal or SharedSignal.")
 
-    # import_halo2_advice and import_halo2_fixed are ignored.
-
-    def step_type(
-        self: CircuitContext, step_type_context: StepTypeContext
-    ) -> StepTypeContext:
-        self.circuit.add_step_type(
-            step_type_context.step_type, step_type_context.step_type.name
+    def step_type(self: Circuit, step_type: StepType) -> StepType:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.add_step_type(step_type.step_type, step_type.step_type.name)
+        return step_type
+
+    def step_type_def(self: StepType) -> StepType:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.add_step_type_def()
+
+    def fixed_gen(self: Circuit, fixed_gen_def: Callable[[FixedGenContext], None]):
+        self.ast.set_fixed_gen(fixed_gen_def)
+
+    def pragma_first_step(self: Circuit, step_type: StepType) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.first_step = step_type.step_type.id
+        print(f"first step id: {step_type.step_type.id}")
+
+    def pragma_last_step(self: Circuit, step_type: StepType) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.last_step = step_type.step_type.id
+        print(f"last step id: {step_type.step_type.id}")
+
+    def pragma_num_steps(self: Circuit, num_steps: int) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.num_steps = num_steps
+
+    def pragma_disable_q_enable(self: Circuit) -> None:
+        assert self.mode == CircuitMode.SETUP
+        self.ast.q_enable = False
+
+    def gen_witness(self: Circuit, args: Any) -> TraceWitness:
+        self.mode = CircuitMode.Trace
+        self.trace_context = TraceContext()
+        self.ast.set_trace(self.trace)
+        self.trace(args)
```

Comment:
I think we should delete self.trace_context before returning, to avoid the idea that we have one witness per circuit. Methods that need a witness, it should be passed as a parameter

Reply:
Agreed and fixed.